### PR TITLE
MySQL compatibility: `toDayOfWeek()` now supports string arguments

### LINKED ISSUE #54794: MySQL compatibility: Illegal type String of argument of function toDayOfWeek
Required for Tableau Online. 

Sample rendered query:

```sql
SELECT FLOOR(((((7 + DAYOFYEAR(cell_towers.updated)) - 1) +
               DAYOFWEEK(DATE_FORMAT(cell_towers.updated, '%Y-01-01 00:00:00'))) - 1) / 7) AS wk_updated_ok
FROM cell_towers
GROUP BY 1;
```

fails with

```
DB::Exception: Illegal type String of argument of function toDayOfWeek. Must be Date, Date32, DateTime or DateTime64.: In scope SELECT FLOOR(((((7 + DAYOFYEAR(cell_towers.updated)) - 1) + DAYOFWEEK(DATE_FORMAT(cell_towers.updated, '%Y-01-01 00:00:00'))) - 1) / 7) AS wk_updated_ok FROM cell_towers GROUP BY 1. (ILLEGAL_TYPE_OF_ARGUMENT)
```

**How to reproduce**

* Which ClickHouse server version to use: latest master
* Which interface to use, if matters: MySQL
* Sample data for all these tables: cell_towers sample dataset

**Expected behavior**

DAYOFWEEK accepts String argument like in MySQL.

### PR DESCRIPTION
**Pull request by rschu1ze**
Largely based on #55572 (kudos to @slvrtrn) but solves the task a) a bit more elegantly + b) also for `toWeek()` and `toYearWeek()`

Resolves #54794

(refactorings are contained in a separate commit, it is easier to review https://github.com/ClickHouse/ClickHouse/pull/55589/commits individually)

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Functions `toDayOfWeek()` (MySQL alias: `DAYOFWEEK()`), `toYearWeek()` (`YEARWEEK()`) and `toWeek()` (`WEEK()`) now supports `String` arguments. This makes its behavior consistent with MySQL's behavior.

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit bb0ff98f5ed12e4ea9a603dd31e2a09e2011341d with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55589/bb0ff98f5ed12e4ea9a603dd31e2a09e2011341d/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6507151745">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55589/bb0ff98f5ed12e4ea9a603dd31e2a09e2011341d/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: docs/en/sql-reference/functions/date-time-functions.md [hunk 1]
```diff
@@ -601,6 +601,8 @@ toDayOfWeek(t[, mode[, timezone]])
 - `mode` - determines what the first day of the week is. Possible values are 0, 1, 2 or 3. See the table above for the differences.
 - `timezone` - optional parameter, it behaves like any other conversion function
 
+The first argument can also be specified as [String](../data-types/string.md) in a format supported by [parseDateTime64BestEffort()](type-conversion-functions.md#parsedatetime64besteffort). This was implemented for reasons of compatibility with MySQL which certain 3rd party tools expect. Because string parsing is slow, it is generally recommended to not use string arguments.
```
Comment (slvrtrn) [inline docs/en/sql-reference/functions/date-time-functions.md:604]: I specifically wanted to emphasize _"do not use on purpose"_ and _"avoid using outside of MySQL interface"_ recommendations in the original doc entry, as we might want to consider limiting the availability of `toDayOfWeek(String)` overload to MySQL interface only as part of making MySQL interface internal implementation more tidy and less invasive.

**Comment by slvrtrn** [comment]
> Function toDayOfWeek() (MySQL alias: DAYOFWEEK()) now supports String arguments. This makes its behavior consistent with MySQL's behavior.

The changelog/PR title needs to be updated with `toWeek()` and `toYearWeek()` as well

**Comment by rschu1ze** [comment]
> The changelog/PR title needs to be updated with `toWeek()` and `toYearWeek()` as well

Done, tnx.

File: docs/en/sql-reference/functions/date-time-functions.md (hunk 1 above)
Comment (rschu1ze) [reply docs/en/sql-reference/functions/date-time-functions.md:604]: Okay, I'll rephrase the note to be more repulsive.

File: src/Functions/CustomWeekTransforms.h [hunk 2]
```diff
@@ -195,7 +70,29 @@ struct CustomWeekTransformImpl
 
         const DateLUTImpl & time_zone = extractTimeZoneFromFunctionArguments(arguments, 2, 0);
         const ColumnPtr source_col = arguments[0].column;
-        if (const auto * sources = checkAndGetColumn<typename FromDataType::ColumnType>(source_col.get()))
+
+        if constexpr (std::is_same_v<FromDataType, DataTypeString>)
+        {
+            /// TODO: remove this entire if constexpr branch, move the parsing code into DateTimeTransforms (yes, even if that will
+            /// duplicate a few lines of code)
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Functions/CustomWeekTransforms.h:77]: Ah, I see ... in that case, the code can stay in place.

**Review by slvrtrn — APPROVED** [review]

**Pull request merged**